### PR TITLE
GraphQL cart items price & currency updated #29041 Solution

### LINKED ISSUE #29041: GraphQL cart items price currency
The currency of the Money object in the cart query are not corresponding to the either Content-Currency and base currency. The value of the price is updated but the currency is not.

### Preconditions (*)

1. Magento 2.4-develop

### Steps to reproduce (*)

1. Setup a Magento store with multiple currencies on the default store view
2. Create cart (graphql)
3. Add product to cart (graphql)
4. Get cart (graphql) - make sure to add non default currency code as Header: Content-Currency (USD is used in example)

```
query {
  cart(cart_id: "nCI33c1lc0JUETLzvQvliMLt9Lpy7URy") {
    items {
      prices {
        row_total {
          value
          currency
        }
      }
    }
  }
}
```

### Expected result (*)

1. Currency property on Money object should return the chosen currency
```
{
    "data": {
        "cart": {
            "items": [
                {
                    "prices": {
                        "price": {
                            "value": 31.48,
                            "currency": "USD"
                        }
                    }
                }
            ]
        }
    }
}
```

### Actual result (*)

1. Currency will always be the default currency
```
{
    "data": {
        "cart": {
            "items": [
                {
                    "prices": {
                        "price": {
                            "value": 31.48,          <- this value is in US dollar
                            "currency": "EUR"        <- non corresponding currency
                        }
                    }
                }
            ]
        }
    }
}
```

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [x] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by makwanamahesh-code**
### Description (*)

1. Fixes magento/magento2#29041 

### Manual testing scenarios (*)
After add above changes out put be like this 
1. Content-Currency: USD
```
{
    "data": {
        "cart": {
            "items": [
                {
                    "prices": {
                        "price": {
                            "value": 9.21,
                            "currency": "USD"
                        }
                    }
                }
            ]
        }
    }
}
```
2. Content-Currency: EUR
```
{
    "data": {
        "cart": {
            "items": [
                {
                    "prices": {
                        "price": {
                            "value": 7.51,
                            "currency": "EUR"
                        }
                    }
                }
            ]
        }
    }
}
```

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: app/code/Magento/QuoteGraphQl/Model/Resolver/CartItemPrices.php [hunk 1]
```diff
@@ -106,4 +129,22 @@ private function getDiscountValues($cartItem, $currencyCode)
...
+    {
+        return $this->storeManager->getStore()->getCurrentCurrencyCode();
+    }
+
+    public function setConvertPrice($price)
+    {
+        $storeId = $this->storeManager->getStore()->getStoreId();
+        $currentCurrencyCode = $this->getCurrentCurrency();
+        $baseCurrencyCode = $this->storeManager->getStore()->getBaseCurrencyCode();
+        if ($baseCurrencyCode == $currentCurrencyCode) {
```
Comment (Stepa4man) [inline app/code/Magento/QuoteGraphQl/Model/Resolver/CartItemPrices.php]: Please use a strict comparison operator.

File: app/code/Magento/QuoteGraphQl/Model/Resolver/CartItemPrices.php [hunk 2]
```diff
@@ -106,4 +129,22 @@ private function getDiscountValues($cartItem, $currencyCode)
...
+            return $price;
+        } else {
+            $price = round($this->priceCurrencyObj->convert($price, $storeId, $currentCurrencyCode), 2);
+            return $price;
```
Comment (Stepa4man) [inline app/code/Magento/QuoteGraphQl/Model/Resolver/CartItemPrices.php]: Could be simplified.

File: app/code/Magento/QuoteGraphQl/Model/Resolver/CartItemPrices.php [hunk 3]
```diff
@@ -106,4 +129,22 @@ private function getDiscountValues($cartItem, $currencyCode)
         }
         return null;
     }
+
+    private function getCurrentCurrency()
+    {
+        return $this->storeManager->getStore()->getCurrentCurrencyCode();
+    }
+
+    public function setConvertPrice($price)
```
Comment (Stepa4man) [inline app/code/Magento/QuoteGraphQl/Model/Resolver/CartItemPrices.php:144]: Please, add a dock block

File: app/code/Magento/QuoteGraphQl/Model/Resolver/CartItemPrices.php [hunk 4]
```diff
@@ -106,4 +129,22 @@ private function getDiscountValues($cartItem, $currencyCode)
         }
         return null;
     }
+
+    private function getCurrentCurrency()
```
Comment (Stepa4man) [inline app/code/Magento/QuoteGraphQl/Model/Resolver/CartItemPrices.php:134]: Please, add a dock block

File: app/code/Magento/QuoteGraphQl/Model/Resolver/CartItemPrices.php (hunk 3 above)
Comment (Stepa4man) [reply app/code/Magento/QuoteGraphQl/Model/Resolver/CartItemPrices.php:144]: Please declare types for params and return.

File: app/code/Magento/QuoteGraphQl/Model/Resolver/CartItemPrices.php (hunk 4 above)
Comment (Stepa4man) [reply app/code/Magento/QuoteGraphQl/Model/Resolver/CartItemPrices.php:134]: Please declare types for params and return.

File: app/code/Magento/QuoteGraphQl/Model/Resolver/CartItemPrices.php [hunk 5]
```diff
@@ -55,12 +74,16 @@ public function resolve(Field $field, $context, ResolveInfo $info, array $value
             // But the totals should be calculated even if no address is set
             $this->totals = $this->totalsCollector->collectQuoteTotals($cartItem->getQuote());
         }
-        $currencyCode = $cartItem->getQuote()->getQuoteCurrencyCode();
+        // update currency code
+        //$currencyCode = $cartItem->getQuote()->getQuoteCurrencyCode();
```
Comment (Stepa4man) [inline app/code/Magento/QuoteGraphQl/Model/Resolver/CartItemPrices.php]: Please don't comment out any code. If we don't need it - delete it

File: app/code/Magento/QuoteGraphQl/Model/Resolver/CartItemPrices.php [hunk 6]
```diff
@@ -30,13 +32,30 @@ class CartItemPrices implements ResolverInterface
...
 
+    /**
+     * @var StoreManager
+     */
+    private $storeManager;
+
+    /**
+     * @var PriceCurrencyObj
+     */
+    private $priceCurrencyObj;
```
Comment (Stepa4man) [inline app/code/Magento/QuoteGraphQl/Model/Resolver/CartItemPrices.php]: Please remove "obj" suffix

File: app/code/Magento/QuoteGraphQl/Model/Resolver/CartItemPrices.php [hunk 7]
```diff
@@ -30,13 +32,30 @@ class CartItemPrices implements ResolverInterface
...
      */
     public function __construct(
-        TotalsCollector $totalsCollector
+        TotalsCollector $totalsCollector,
+        StoreManagerInterface $storeManager,
+        PriceCurrencyInterface $priceCurrencyObj
     ) {
         $this->totalsCollector = $totalsCollector;
+        $this->storeManager = $storeManager;
+        $this->priceCurrencyObj = $priceCurrencyObj;
```
Comment (Stepa4man) [inline app/code/Magento/QuoteGraphQl/Model/Resolver/CartItemPrices.php]: Please remove "obj" suffix

**Review by Stepa4man — CHANGES_REQUESTED** [review]
Hi @maheshmakwana-code 
Thanks for your contribution.

Please take a look at my comments.

**Comment by maheshmakwana-code** [comment]
Please help me to do  what and which command and how to to contribute here  ?

**Comment by maheshmakwana-code** [comment]
@magento run all tests

**Comment by maheshmakwana-code** [comment]
@magento run Database Compare, Database Compare, Functional Tests CE, Functional Tests EE,Integration Tests, Magento Health Index, Sample Data Tests CE,Sample Data Tests EE, Sample Data Tests B2B, Static Tests, Unit Tests ,WebAPI Tests

**Comment by maheshmakwana-code** [comment]
there are only one Functional Tests B2B  is Failing. My solution is only for open source edition. so now what i do ?

**Comment by maheshmakwana-code** [comment]
@magento run all tests

**Comment by maheshmakwana-code** [comment]
@magento run Unit Tests

**Comment by maheshmakwana-code** [comment]
Why Merging is blocked ?

**Comment by maheshmakwana-code** [comment]
> Hi @maheshmakwana-code
> Thanks for your contribution.
> 
> Please take a look at my comments.

Please guide and help me to contribution.

**Comment by maheshmakwana-code** [comment]
@magento run all tests

**Comment by maheshmakwana-code** [comment]
magento run all tests

**Comment by maheshmakwana-code** [comment]
@magento run all tests

**Comment by maheshmakwana-code** [comment]
@magento run Functional Tests CE

**Comment by maheshmakwana-code** [comment]
**The base branch restricts merging to authorized users.**


@magento  Please help me to solve this

**Comment by maheshmakwana-code** [comment]
@magento run all tests

**Comment by maheshmakwana-code** [comment]
@magento run Functional Tests CE

**Comment by maheshmakwana-code** [comment]
@magento run Functional Tests CE

**Comment by maheshmakwana-code** [comment]
@magento run all tests

**Comment by maheshmakwana-code** [comment]
@magento run Functional Tests CE ,Functional Tests EE

**Comment by maheshmakwana-code** [comment]
@magento run all tests

**Comment by maheshmakwana-code** [comment]
@magento Functional Tests EE,

**Comment by maheshmakwana-code** [comment]
@magento run Functional Tests EE

**Comment by maheshmakwana-code** [comment]
@magento run Functional Tests EE

File: app/code/Magento/QuoteGraphQl/Model/Resolver/CartItemPrices.php [hunk 8]
```diff
@@ -106,4 +126,35 @@ private function getDiscountValues($cartItem, $currencyCode)
...
+        return $this->storeManager->getStore()->getCurrentCurrencyCode();
+    }
+
+    /**
+     * Get current currency price
+     *
+     * @param string $price
+     * @return string
+     */
+    public function setConvertPrice($price)
```
Comment (Stepa4man) [inline app/code/Magento/QuoteGraphQl/Model/Resolver/CartItemPrices.php:144]: Please declare types for params and return.

File: app/code/Magento/QuoteGraphQl/Model/Resolver/CartItemPrices.php [hunk 9]
```diff
@@ -106,4 +126,35 @@ private function getDiscountValues($cartItem, $currencyCode)
         }
         return null;
     }
+
+    /**
+     * Get Current Currency
+     *
+     * @return string
+     */
+
```
Comment (Stepa4man) [inline app/code/Magento/QuoteGraphQl/Model/Resolver/CartItemPrices.php:133]: Please remove empty line

File: app/code/Magento/QuoteGraphQl/Model/Resolver/CartItemPrices.php [hunk 10]
```diff
@@ -106,4 +126,35 @@ private function getDiscountValues($cartItem, $currencyCode)
...
+    private function getCurrentCurrency()
```
Comment (Stepa4man) [inline app/code/Magento/QuoteGraphQl/Model/Resolver/CartItemPrices.php:134]: Please declare return type.

File: app/code/Magento/QuoteGraphQl/Model/Resolver/CartItemPrices.php [hunk 11]
```diff
@@ -106,4 +126,35 @@ private function getDiscountValues($cartItem, $currencyCode)
...
+    {
+        $storeId = $this->storeManager->getStore()->getStoreId();
+        $currentCurrencyCode = $this->getCurrentCurrency();
+        $baseCurrencyCode = $this->storeManager->getStore()->getBaseCurrencyCode();
+        if ($baseCurrencyCode === $currentCurrencyCode) {
+            return $price;
+        } else {
+            $price = $this->priceCurrency->convert($price, $storeId, $currentCurrencyCode);
+            $price = round($price , 2);
+            return $price;
```
Comment (Stepa4man) [inline app/code/Magento/QuoteGraphQl/Model/Resolver/CartItemPrices.php]: Can be simplified

File: app/code/Magento/QuoteGraphQl/Model/Resolver/CartItemPrices.php [hunk 12]
```diff
@@ -106,4 +126,35 @@ private function getDiscountValues($cartItem, $currencyCode)
...
+     * @return string
+     */
+    public function setConvertPrice($price)
+    {
+        $storeId = $this->storeManager->getStore()->getStoreId();
+        $currentCurrencyCode = $this->getCurrentCurrency();
+        $baseCurrencyCode = $this->storeManager->getStore()->getBaseCurrencyCode();
+        if ($baseCurrencyCode === $currentCurrencyCode) {
+            return $price;
+        } else {
```
Comment (Stepa4man) [inline app/code/Magento/QuoteGraphQl/Model/Resolver/CartItemPrices.php]: Why do you need `else` here?

**Review by Stepa4man — CHANGES_REQUESTED** [review]
Hi @maheshmakwana-code 
Please take a look at my comments. If you need help with params and return type declaration, please rich me in slack or ask here.

**Comment by maheshmakwana-code** [comment]
@magento run all tests

**Comment by maheshmakwana-code** [comment]
@magento run Functional Tests EE, Functional Tests CE, Functional Tests B2B

**Comment by maheshmakwana-code** [comment]
@magento run Functional Tests EE, Functional Tests CE, Functional Tests B2B

File: app/code/Magento/QuoteGraphQl/Model/Resolver/CartItemPrices.php (hunk 8 above)
Comment (makwanamahesh-code) [reply app/code/Magento/QuoteGraphQl/Model/Resolver/CartItemPrices.php:144]: How to rich you in Slack ?

**Comment by maheshmakwana-code** [comment]
@magento run all tests

**Comment by maheshmakwana-code** [comment]
@magento run all tests

**Review by makwanamahesh-code — COMMENTED** [review]
I have done your changes and how to contact to you sir?

**Comment by maheshmakwana-code** [comment]
@magento run all tests

**Comment by maheshmakwana-code** [comment]
@magento run Unit Tests, Functional Tests CE, Functional Tests EE, Functional Tests B2B, Integration Tests

**Comment by maheshmakwana-code** [comment]
@magento run Functional Tests CE, Functional Tests EE, Functional Tests B2B

**Comment by maheshmakwana-code** [comment]
@magento run Functional Tests CE

**Comment by maheshmakwana-code** [comment]
@magento run Functional Tests EE

**Comment by maheshmakwana-code** [comment]
@magento run all tests

**Review by Stepa4man — CHANGES_REQUESTED** [review]
@maheshmakwana-code
Please don't "resolve" comments if you didn't fix it or at least discussed it.

Please review them again and merge the latest 2.4-develop branch in your branch. Thanks.

**Comment by maheshmakwana-code** [comment]
@magento run all tests

**Comment by maheshmakwana-code** [comment]
@magento run all tests

**Comment by maheshmakwana-code** [comment]
@magento run all tests

**Comment by Stepa4man** [comment]
I close this ticket due to inactivity.

**Pull request closed without merging**